### PR TITLE
Add CIAM sample

### PR DESCRIPTION
**Pull request by Avery-Dunn**
Adds a Java servlet sample showing how to use MSAL Java to sign a user into a CIAM tenant in a web app, as per https://github.com/AzureAD/microsoft-authentication-library-for-java/issues/724 . A web API sample will be added soon.

Also updates the samples the use the latest MSAL Java version, plus some other minor fixes.

### CODE REVIEW

File: 1-server-side/msal-client-credential-secret/src/main/java/ClientCredentialGrant.java [hunk 1]
```diff
@@ -62,7 +62,7 @@ public static void main(String args[]) throws Exception{
 
     private static void GetOrCreateApp(String clientId, String secret, String authority) throws MalformedURLException {
 
-        if (app!=null) {
+        if (app == null) {
```
Comment (bgavrilMS) [inline 1-server-side/msal-client-credential-secret/src/main/java/ClientCredentialGrant.java:65]: Oh snap, I wonder who introduced this :(

File: 3-java-servlet-web-app/1-Authentication/sign-in-ciam/README.md [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+
+1. This web application uses **MSAL for Java (MSAL4J)** to sign in users to their own Microsoft Entra tenant and obtains an [ID Token](https://docs.microsoft.com/azure/active-directory/develop/id-tokens) from **Microsoft Entra ID**.
+2. The **ID Token** proves that a user has successfully authenticated with this tenant.
+3. The web application protects one of its routes according to user's authentication status.
+
+## Contents
+
+| File/folder                                                        | Description                                                                            |
+| ------------------------------------------------------------------ | -------------------------------------------------------------------------------------- |
+| `AppCreationScripts/`                                              | Scripts to automatically configure Microsoft Entra app registrations.                         |
```
Comment (bgavrilMS) [inline 3-java-servlet-web-app/1-Authentication/sign-in-ciam/README.md:52]: Do AppCreationScripts really work for CIAM? If unsure, let's leave them out

File: 3-java-servlet-web-app/1-Authentication/sign-in-ciam/README.md [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+## About the code
+
+This sample shows how to use **MSAL for Java (MSAL4J)** to sign in users into your Microsoft Entra tenant. If you'd like to use MSAL4J in your own applications, must add it to your projects using Maven.
+
+If you want to replicate this sample's behavior, you may choose to copy the `pom.xml` file, and the contents of the `helpers` and `authservlets` packages in the `src/main/java/com/microsoft/azuresamples/msal4j` package. You'll also need the [authentication.properties file](src/main/resources/authentication.properties). These classes and files contain generic code that can be used in a wide array of applications. The rest of the sample may be copied as well, but the other classes and files are built specifically to address this sample's objective.
+
+A **ConfidentialClientApplication** instance is created in the [AuthHelper.java](src/main/java/com/microsoft/azuresamples/authentication/AuthHelper.java) class. This object helps craft the Microsoft Entra authorization URL and also helps exchange the authentication token for an access token.
+
+```Java
+// getConfidentialClientInstance method
```
Comment (bgavrilMS) [inline 3-java-servlet-web-app/1-Authentication/sign-in-ciam/README.md:157]: dead code ?

File: 3-java-servlet-web-app/1-Authentication/sign-in-ciam/README.md [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+                    .build();
+```
+
+The following parameters need to be provided upon instantiation:
+
+- The **Client ID** of the app
+- The **Client Secret**, which is a requirement for Confidential Client Applications
+- The **Microsoft Entra authority**, which includes your Microsoft Entra tenant ID.
+
+In this sample, these values are read from the [authentication.properties](src/main/resources/authentication.properties) file using a properties reader in the class [Config.java](src/main/java/com/microsoft/azuresamples/authentication/Config.java).
```
Comment (bgavrilMS) [inline 3-java-servlet-web-app/1-Authentication/sign-in-ciam/README.md:171]:
```suggestion
In this sample, these values are read from the [authentication.properties](src/main/resources/authentication.properties) file using a properties reader in the class [Config.java](src/main/java/com/microsoft/azuresamples/authentication/Config.java). However, in production, the `SECRET` must be stored in a secure location (e.g. Azure KeyVault). Do not store secrets in code. 
```

File: 3-java-servlet-web-app/1-Authentication/sign-in-ciam/README.md [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+
+### Step-by-step walkthrough
+
+1. The first step of the sign-in process is to send a request to the `/authorize` endpoint on for our Microsoft Entra tenant. Our MSAL4J `ConfidentialClientApplication` instance is leveraged to construct an authorization request URL. Our app redirects the browser to this URL, which is where the user will sign in.
+
+    ```Java
+    final ConfidentialClientApplication client = getConfidentialClientInstance();
+    AuthorizationRequestUrlParameters parameters = AuthorizationRequestUrlParameters.builder(Config.REDIRECT_URI, Collections.singleton(Config.SCOPES))
+            .responseMode(ResponseMode.QUERY).prompt(Prompt.SELECT_ACCOUNT).state(state).nonce(nonce).build();
```
Comment (bgavrilMS) [inline 3-java-servlet-web-app/1-Authentication/sign-in-ciam/README.md:180]: Is "SELECT_ACCOUNT" prompt needed here? Does the token issuer's default value not provide a good experience?

File: 3-java-servlet-web-app/1-Authentication/sign-in-ciam/src/main/java/com/microsoft/azuresamples/msal4j/authservlets/AADRedirectServlet.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    @Override
+    protected void doGet(final HttpServletRequest req, final HttpServletResponse resp)
+            throws ServletException, IOException {
+        logger.log(Level.FINE, "Request has come with params {0}", req.getQueryString());
+        try {
+            AuthHelper.processAADCallback(new IdentityContextAdapterServlet(req, resp));
+            logger.log(Level.INFO, "redirecting to home page.");
+            resp.sendRedirect(Config.HOME_PAGE);
+        } catch (AuthException ex) {
+            logger.log(Level.WARNING, ex.getMessage());
```
Comment (bgavrilMS) [inline 3-java-servlet-web-app/1-Authentication/sign-in-ciam/src/main/java/com/microsoft/azuresamples/msal4j/authservlets/AADRedirectServlet.java:39]: nit: use ERROR level ?
nit: isn't there an overload of `log` that deals with exceptions?

File: 3-java-servlet-web-app/1-Authentication/sign-in-ciam/src/main/java/com/microsoft/azuresamples/msal4j/helpers/Config.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+                Config.logger.log(Level.SEVERE, "FATAL: Could not load required key {0} from config! EXITING", key);
+                System.exit(1); // HANDLE THIS BETTER IN YOUR APP.
+                return null;
+            } else {
+                Config.logger.log(Level.WARNING, "Could not load {0}!", key);
+                return "";
+            }
+        } else {
+            Config.logger.log(Level.SEVERE, "FATAL: Could not load property reader! EXITING!");
+            System.exit(1);
```
Comment (bgavrilMS) [inline 3-java-servlet-web-app/1-Authentication/sign-in-ciam/src/main/java/com/microsoft/azuresamples/msal4j/helpers/Config.java:73]: Seems strange. Why not use exceptions?

File: 3-java-servlet-web-app/1-Authentication/sign-in-ciam/src/main/java/com/microsoft/azuresamples/msal4j/helpers/GraphHelper.java [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+import com.microsoft.graph.requests.GroupCollectionPage;
+import com.microsoft.graph.requests.GroupCollectionRequest;
+import com.microsoft.graph.requests.GroupCollectionRequestBuilder;
+import com.microsoft.graph.models.Group;
+
+/**
+ * GraphHelper class handles creating a Graph SDK client (IGraphServiceClient)
+ * and has functions for common Microsoft Graph calls
+ */
+public class GraphHelper {
```
Comment (bgavrilMS) [inline 3-java-servlet-web-app/1-Authentication/sign-in-ciam/src/main/java/com/microsoft/azuresamples/msal4j/helpers/GraphHelper.java:28]: This sample doesn't call Graph. Why is this here?

**Review by bgavrilMS — APPROVED** [review]

File: 1-server-side/msal-client-credential-secret/src/main/java/ClientCredentialGrant.java (hunk 1 above)
Comment (Avery-Dunn) [reply 1-server-side/msal-client-credential-secret/src/main/java/ClientCredentialGrant.java:65]: Yeah, just noticed that when testing CIAM with other flows (the other places you added this method are correct though!)

File: 3-java-servlet-web-app/1-Authentication/sign-in-ciam/README.md [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+- [Code of Conduct](#code-of-conduct)
+- [More information](#more-information)
+
+## Overview
+
+This sample demonstrates a Java Servlet web app that signs in users to your Microsoft Entra CIAM tenant using the [Microsoft Authentication Library (MSAL) for Java](https://github.com/AzureAD/microsoft-authentication-library-for-java).
+
+## Scenario
+
+1. This web application uses **MSAL for Java (MSAL4J)** to sign users into a Microsoft Entra CAIM tenant and obtain an [ID Token](https://docs.microsoft.com/azure/active-directory/develop/id-tokens) from **Microsoft Entra ID**.
```
Comment (DidunAyodeji) [inline 3-java-servlet-web-app/1-Authentication/sign-in-ciam/README.md:44]: minor typo. "CAIM tenant" instead of CIAM

Comment (Avery-Dunn) [reply 3-java-servlet-web-app/1-Authentication/sign-in-ciam/README.md]: Good catch, fixed in the latest commit

File: 3-java-servlet-web-app/1-Authentication/sign-in-ciam/src/main/java/com/microsoft/azuresamples/msal4j/helpers/GraphHelper.java (hunk 8 above)
Comment (Avery-Dunn) [reply 3-java-servlet-web-app/1-Authentication/sign-in-ciam/src/main/java/com/microsoft/azuresamples/msal4j/helpers/GraphHelper.java]: I originally copied most of the code from the basic "sign-in" sample, and if you look [here](https://github.com/Azure-Samples/ms-identity-msal-java-samples/blob/0f338b0d20fd8d1e540bd24b1fee2f300643a0b9/3-java-servlet-web-app/1-Authentication/sign-in/src/main/java/com/microsoft/azuresamples/msal4j/helpers/AuthHelper.java#L92) you can see how it could be used: apparently just as a way to handle a user in too many groups.

It might be helpful for the samples that deal with groups but isn't relevant here, so in the latest commit I just deleted this class and it's potential references.

File: 3-java-servlet-web-app/1-Authentication/sign-in-ciam/src/main/java/com/microsoft/azuresamples/msal4j/helpers/Config.java (hunk 7 above)
Comment (Avery-Dunn) [reply 3-java-servlet-web-app/1-Authentication/sign-in-ciam/src/main/java/com/microsoft/azuresamples/msal4j/helpers/Config.java]: I had just copied this from another servlet sample. It's definitely strange, maybe the idea was that if the properties couldn't load there was no point in even running the program?

Regardless, in the latest commit I replaced the system.exit with exceptions and removed some other unneeded code.

File: 3-java-servlet-web-app/1-Authentication/sign-in-ciam/README.md (hunk 5 above)
Comment (Avery-Dunn) [reply 3-java-servlet-web-app/1-Authentication/sign-in-ciam/README.md:186]: This was copied from the basic sign-in sample, and after trying with and without it definitely doesn't seem needed. Removed in the latest commit, and I'll be going through the other samples too at some point.

File: 3-java-servlet-web-app/1-Authentication/sign-in-ciam/README.md (hunk 3 above)
Comment (Avery-Dunn) [reply 3-java-servlet-web-app/1-Authentication/sign-in-ciam/README.md]: This was just a comment saying that the code is a snippet of the `getConfidentialClientInstance` method. The method itself wasn't much longer than the snippet, so in the latest commit I just added the whole method to this section.

File: 3-java-servlet-web-app/1-Authentication/sign-in-ciam/README.md (hunk 2 above)
Comment (Avery-Dunn) [reply 3-java-servlet-web-app/1-Authentication/sign-in-ciam/README.md]: The existing scripts in other samples wouldn't have worked for this one so I didn't copy the code, but I forgot to remove this reference to it. Fixed in the latest commit.
